---
title: "Ecoli-Human Spike-in TMT Study (PRIDE ID  PXD013277)"
author:
- name: Lieven Clement, Oliver M. Crook
output:
  BiocStyle::html_document
---

*Disclaimer*: we only have a workflow for labelled proteomics when all samples were quantified within the same run. We are currently developing bespoke workflows for labelled proteomics experiments with multiple runs.

# Background
In this vignette, we assess the performance of MSqRob for differential
expression analysis using a publicly available spike-in study
(PRIDE identifier: PXD013277).
One MCF-7 and one E.coli K-12 cell pellet were lysed and
sonicated in a buffer containing 4% SDS, 25 mM HEPES pH 7.6 and 1mM DTT. Total
protein amount was estimated (Bio-Rad DC). Samples with different spiked in amounts of E.coli protein extract (3 replicates with 7.5 µg, 4 with 15 µg and 3 with 45 µg) in MCF-7 background (70 µg of protein extract) were prepared. Protein digestion (LysC and trypsin, sequencing grade modified, Pierce) was performed using a modified SP3-protocol.

Spectra data was converted to mzML files using ProteoWizard release: 3.0.10827 (2017-5-11) and searched with MS-GF+ (2016.10.26) and Percolator. Precursor mass tolerance used was 10 ppm, fragment mass tolerance 0.11 Da, fixed modifications
were TMT-10plex on lysines and peptide N-termini, and carbamidomethylation on cysteine residues, oxidation on methionine was used as a variable modification. The
protein database used for search was Uniprot (2018_04) human protein databases with E.coli protein database concatenated (78807 protein sequences) allowing for one tryptic miss-cleavage. PSMs and proteins were filtered at 1% FDR resulting in 308,001 PSMs; 122,235 unique peptides and 11,216 proteins.


# Data
```{r}
library(msqrob2)
url <- "ftp://jpost.pharm.kyoto-u.ac.jp/repository/JPST000562/target_psmtable.txt"
download.file(url, destfile = "./target_psmtable.txt",method = "auto")

df <- read.table("target_psmtable.txt",stringsAsFactors = FALSE, header = TRUE, quote = "", comment.char = "",sep = "\t")
```

We make a unique name per psm and read in data.
The intensities are in the columns starting with the string "tmt10plex".

```{r}
ecols<-grep("tmt10plex",colnames(df))
pe <- readQFeatures(df,ecol=ecols,name="psmRaw")

colData(pe)$condition<-as.factor(c(rep("a",3),rep("b",4),rep("c",3)))
rm(df)
gc()
```


We calculate how many non zero intensities we have per peptide.
This will be useful for filtering.

```{r}
rowData(pe[["psmRaw"]])$nNonZero <- rowSums(assay(pe[["psmRaw"]]) > 0)
```


Peptides with zero intensities are missing peptides and should be
represent with a `NA` value instead of `0`.
```{r}
pe <- zeroIsNA(pe, i = "psmRaw")
```

# Preprocessing

## Log transform the data

```{r}
pe <- logTransform(pe, base = 2, i = "psmRaw", name = "psmLog")
colPal <- colorRampPalette(c("red", "green", "blue"),
                                     space = "rgb")
limma::plotDensities(assay(pe[["psmLog"]]),col=colPal(10))
```

## Filtering

We will retain the psms that are picked up in at least four channels

```{r}
filtObserved <- VariableFilter(field = "nNonZero",
                               value =4,
                               condition =">=")
pe <- filterFeatures(pe,filtObserved)
```

## Summarization

We will summarize the psms to proteins by using the median polish algorithm.
We do not normalize the psm data first because this would break the connection between the reporter ions.

```{r}
pe <- aggregateFeatures(pe,           
                        i="psmLog",
                        fcol="Master.protein.s.",
                        name="protein",
                        fun=MsCoreUtils::medianPolish,
                        na.rm=TRUE
                        )
limma::plotDensities(assay(pe[["protein"]]),col=colPal(10))
```

## Normalisation

We use a DESeq2 style normalisation to correct for differences in mass loading in the different reporter channels.

$$ s_j = median_i \left[\frac{y_{ij}}{median_k(y_{ik})}\right] $$

On a log scale this becomes an offset:

$$ \log_2 s_j = median_i \left[ \left(\log_2 y_{ij} - median_k(\log_2 y_{ik})\right)\right] $$

```{r}
offset <- colMedians(assay(pe[["protein"]])-rowMedians(assay(pe[["protein"]]),na.rm=TRUE),na.rm=TRUE)
assay(pe[["protein"]]) <- t(t(assay(pe[["protein"]]))-offset)
limma::plotDensities(assay(pe[["protein"]]),col=colPal(10))
```

# Analysis

## Estimation

We model the protein level expression values using msqrob.
By default msqrob2 estimates the model parameters using robust regression.

```{r, warning=FALSE}
pe <- msqrob(object = pe, i = "protein", formula = ~condition)
```

## Inference

What are the parameter names of the model?
```{r}
getCoef(rowData(pe[["protein"]])$msqrobModels[[1]])
```

Spike-in condition a is the reference class. So the mean log2 expression for
samples from condition a is `(Intercept)`.
The mean log2 expression for samples from condition b-e is
'(Intercept)+conditionb',...,'(Intercept)+conditione', respectively.
Hence, the average log2 fold change (FC) between condition b and condition a is modelled using the parameter 'conditionb'. Thus, we assess the contrast
'conditionb = 0' with our statistical test. The same holds for comparison c-a.


```{r}
L <- makeContrast(c("conditionb = 0", "conditionc = 0", "conditionc - conditionb = 0"), parameterNames = c("conditionb","conditionc"))
pe <- hypothesisTest(pe,"protein",L)
```

# Performance

We know the ground truth. All ecoli proteins are DE, human proteins not.

Function to calculate TPR and FDP
```{r}
tprFdp <- function(pval, tp, adjPval){
ord <- order(pval)
return(data.frame(
  pval = pval[ord],
  adjPval = adjPval[ord],
  tpr = cumsum(tp[ord])/sum(tp),
  fdp = cumsum(!tp[ord])/1:length(tp)))
}
```

## comparisons

```{r}
rowData(pe[["protein"]])$ecoli <- grepl("ECOLI",rownames(pe[["protein"]]))
plot(tpr~fdp,tprFdp(rowData(pe[["protein"]])$conditionb$pval,rowData(pe[["protein"]])$ecoli,rowData(pe[["protein"]])$conditionb$adjPval),type="l",xlab="FDP",ylab="TPR")
lines(tpr~fdp,tprFdp(rowData(pe[["protein"]])$conditionc$pval,rowData(pe[["protein"]])$ecoli,rowData(pe[["protein"]])$conditionc$adjPval),type="l",col=2)
lines(tpr~fdp,tprFdp(rowData(pe[["protein"]])[["conditionc - conditionb"]]$pval,rowData(pe[["protein"]])$ecoli,rowData(pe[["protein"]])[["conditionc - conditionb"]]$adjPval),type="l",col=3)
legend("bottomright",legend=c("B-A","C-A","C-B"),col=1:3,lty=1)
```

## Fold changes

```{r}
boxplot(rowData(pe[["protein"]])[["conditionb"]]$logFC ~ rowData(pe[["protein"]])$ecoli,ylab="log2 FC",xlab="spiked",main="B vs A")
abline(h=log2(2),col=2)
boxplot(rowData(pe[["protein"]])[["conditionc"]]$logFC ~ rowData(pe[["protein"]])$ecoli,ylab="log2 FC",xlab="spiked",main="C vs A")
abline(h=log2(6),col=2)
boxplot(rowData(pe[["protein"]])[["conditionc - conditionb"]]$logFC ~ rowData(pe[["protein"]])$ecoli,ylab="log2 FC",xlab="spiked",main="C vs B")
abline(h=log2(3),col=2)
```

Fold changes are under estimated!

```{r}
rowData(pe[["psmLog"]])$ecoli <- grepl("ECOLI",rowData(pe[["psmLog"]])$Master.protein.s.)
fcBa <- assay(pe[["psmLog"]]) %*% c(rep(-1/3,3),rep(1/4,4),rep(0,3))
fcCa <- assay(pe[["psmLog"]]) %*% c(rep(-1/3,3),rep(0,4),rep(1/3,3))
fcCb <- assay(pe[["psmLog"]]) %*% c(rep(0,3),rep(-1/4,4),rep(1/3,3))
boxplot(fcBa ~ rowData(pe[["psmLog"]])$ecoli,ylab="log2 FC",xlab="spiked",main="B vs A")
abline(h=log2(2),col=2)
boxplot(fcCa ~ rowData(pe[["psmLog"]])$ecoli,ylab="log2 FC",xlab="spiked",main="C vs A")
abline(h=log2(6),col=2)
boxplot(fcCb ~ rowData(pe[["psmLog"]])$ecoli,ylab="log2 FC",xlab="spiked",main="C vs B")
abline(h=log2(2),col=2)
```

# References
